content/backend/typescript-apollo/6-authentication.md

```diff
@@ -0,0 +1,776 @@
+
+
+### Updating the `post` mutation
+
+Now that there is an authentication scheme in place, you are going to make two changes to the `post` mutation. Firstly you are going to enforce only authenticated users can create a new `link` using the `post` mutation. Secondly, in your database, you are going to connect the `link` with the user who is posting through the `postedBy` relation.
+
+<Instruction>
+
+Update the implementation of the `post` resolver in `src/graphql/Link.ts`:
+
+```typescript{11-26}(path="../hackernews-typescript/src/graphql/Link.ts")
+
+export const LinkMutation = extendType({
+    type: "Mutation",
+    definition(t) {
+        t.nonNull.field("post", {
+            type: "Link",
+            args: {
+                description: nonNull(stringArg()),
+                url: nonNull(stringArg()),
+            },
+            resolve(parent, args, context) {   
+                const { description, url } = args;
+                const { userId } = context;
+
+                if (!userId) {  // 1
+                    throw new Error("Cannot post without logging in.");
+                }
+
+                const newLink = context.prisma.link.create({
+                    data: {
+                        description,
+                        url,
+                        postedBy: { connect: { id: userId } },  // 2
+                    },
+                });
+
+                return newLink;
+            },
+        });
+    },
+});
+```
+
+</Instruction>
+
+Let's go through the two changes: 
+
+1. If the `userId` does not exist in `context`, the resolver raises an error. As a result, only authorized users can add a new `link`.  
+
+2. To connect the `User` with the `Link`, you are specifying a value for the `postedBy` field (which represnts this `Link` to `User` relation). The `connect` operator is used by Prisma to specify which `user` the newly created `link` should be associated with. 
+
+At long last, the code for this chapter is complete! This was a pretty long chapter, give yourself a congratulations for finishing it. Now it's time to test the authentication feature. 
+
+### Testing the authentication flow
+
+The very first thing you'll do is test the `signup` mutation and thereby create a new `User` in the database.
+
+<Instruction>
+
+Go to Apollo Studio Explorer at http://localhost:3000/  and send the following mutation to create a new `User`:
+
+```graphql
+mutation {
+  signup(name: "Alice", email: "alice@prisma.io", password: "graphql") {
+    token
+    user {
+      id
+    }
+  }
+}
+```
+
+</Instruction>
+
+<Instruction>
+
+From the server's response, copy the authentication `token`. Now in the `Headers` tab in the bottom middle, add a new header. The name of the header will be `Authorization` and value will be `"Bearer __TOKEN__`". Make sure that the header is enabled by clicking the blue tick mark. 
+
+
+</Instruction>
+
+![Adding Authorization Header](https://i.imgur.com/dFekQoe.png)
```

Comment:
```suggestion
![Adding Authorization Header](https://imgur.com/rB5glDx.png)
```

Reply:
placeholders are also a very difficult thing in technical writing to get right. for beginners. usually I am very explicit about calling out when I'm using a placeholder and tell readers that they need to replace the placeholder with their own value, then providing an example what an actual value might look like.

Reply:
Could you clarify this a bit further? From what I understand, you're saying that: 

1. In the _writing_ I should use a placeholder and make it explicit that it needs to be replaced (This is already being done in the instruction text).
2. In the _example_ (image/screenshot/output) I should use a concrete token/value/etc even though it might differ from the user's token/value. (This is currently not being done in the image and this is what this suggestion is changing). 

Does this make sense?

Reply:
My main point is about making things very explicit, beginner-friendly and idiot-proof. I remember my first programming days where I read tutorials that used something similar to `__TOKEN__` as placeholders and I just did not know that I was supposed to replace it because the tutorial writer made the implicit assumption the readers would understand their notation. So, in an ideal world I think the following notes are made in the tutorial:

- mention explicitly that the code snippet contains a placeholder that is to be replaced
- show the code snippet with the placeholder
- show the code snipped with a sample value explicitly calling out that this is a sample value
- make sure readers understand how they can get their own version of this value that they're supposed to put in place